Classes/common/Encryption/Keychain/CDTEncryptionKeychainManager.m

```diff
@@ -0,0 +1,193 @@
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Number of stored iterations does NOT match the constant value %li",
+                   (long)CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS);
+
+        return NO;
+    }
+
+    // Ensure IV has the correct length
+    if (data.iv.length != CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"IV does not have the expected size: %i bytes",
+                   CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE);
+
+        return NO;
+    }
+
+    return YES;
+}
+
+- (NSData *)generateDpk
+{
+    NSData *dpk = [CDTEncryptionKeychainUtils
+        generateRandomBytesInBufferWithLength:CDTENCRYPTION_KEYCHAIN_ENCRYPTIONKEY_SIZE];
+
+    return dpk;
+}
+
+- (CDTEncryptionKeychainData *)keychainDataToStoreDpk:(NSData *)dpk
+                                encryptedWithPassword:(NSString *)password
+{
+    NSData *salt = [self generatePBKDF2Salt];
+
+    NSData *key =
+        [self generateAESKeyUsingPBKDF2ToDerivePassword:password
+                                               withSalt:salt
+                                             iterations:CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS
+                                                 length:CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE];
+
+    NSData *iv = [self generateAESIv];
+
+    NSData *encryptedDpk = [self encryptDpk:dpk usingAESWithKey:key iv:iv];
+
+    CDTEncryptionKeychainData *keychainData =
+        [CDTEncryptionKeychainData dataWithEncryptedDPK:encryptedDpk
+                                                   salt:salt
+                                                     iv:iv
+                                             iterations:CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS
+                                                version:CDTENCRYPTION_KEYCHAIN_VERSION];
+
+    return keychainData;
+}
+
+- (NSData *)generatePBKDF2Salt
+{
+    NSData *salt = [CDTEncryptionKeychainUtils
+        generateRandomBytesInBufferWithLength:CDTENCRYPTION_KEYCHAIN_PBKDF2_SALT_SIZE];
+
+    return salt;
+}
+
+- (NSData *)generateAESKeyUsingPBKDF2ToDerivePassword:(NSString *)password
+                                             withSalt:(NSData *)salt
+                                           iterations:(NSInteger)iterations
+                                               length:(NSUInteger)length
+{
+    NSData *aesKey = [CDTEncryptionKeychainUtils generateKeyOfLength:length
+                                                       usingPassword:password
+                                                            withSalt:salt
+                                                          iterations:iterations];
+
+    return aesKey;
+}
+
+- (NSData *)generateAESIv
+{
+    NSData *iv = [CDTEncryptionKeychainUtils
+        generateRandomBytesInBufferWithLength:CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE];
+
+    return iv;
+}
+
+- (NSData *)encryptDpk:(NSData *)dpk usingAESWithKey:(NSData *)key iv:(NSData *)iv
```

Comment:
As mentioned in the [previous comment](https://github.com/cloudant/CDTDatastore/pull/128#discussion_r29869733), I'd rather change the name of the methods in `CDTEncryptionKeychainUtils `.